index.rst

```diff
@@ -1,61 +1,499 @@
+
+This interface does not provide enough functionality for most provenance queries, however - it just finds all quanta matching certain criteria, regardless of their relationships - so it is best considered way to obtain a starting point.
+For those, we envision an operation that starts with a set of quanta and traverses the graph according to certain criteria, querying the database as necessary (perhaps once per task or dataset type) and returning matching quanta as it goes::
+
+  def traverseQuanta(
+      self,
+      initial: Iterable[Quantum],
+      forward_while: Optional[str | TraversalPredicate] = None,
+      forward_until: Optional[str | TraversalPredicate] = None,
+      backward_while: Optional[str | TraversalPredicate] = None,
+      backward_until: Optional[str | TraversalPredicate] = None,
+  ) -> Iterable[Quantum]:
+      ...
+
+Traversal could proceed forward (in the same direction as execution) or backward (from outputs back to inputs) or both.
+The criteria for which quanta to traverse and return are encoded in the four predicate arguments, which are *conceptually* just boolean functions on a quantum::
+
+  class TraversalPredicate(ABC):
+
+      @abstractmethod
+      def __call__(self, quantum: Quantum) -> bool:
+          ...
+
+Traversal would be terminated in a direction whenever a ``while`` predicate evaluates to `False` (without returning that quantum) or whenever the corresponding ``until`` predicate evaluates to `True` (which does return that quantum).
+
+This simple conceptual definition of the predicate may not be possible in practice for performance reasons; traversal actually involves database queries, and while we can perform some post-query filtering in Python, we want most of the filtering to happen in the database.
+In practice, then, we may need to define an enumerated library of ``TraversalPredicates``, and perhaps define logical operations to combine them, restricted to what we can translate to SQL queries.
+Most common provenance queries could be satisfied by the following predicates and their negations (even if they cannot be combined):
+
+- whether the quantum's task label is in a given set;
+- whether any input or output dataset type is in given set;
+- whether any input or output dataset UUID is in a given set.
+
+It is worth noting here that the ``Quantum`` and ``QuantumGraph`` objects returned here are not necessarily the same types as those used prior to execution; execution adds more information that we want the provenance system to be able to return.
+Whether to actually use different types involves a lot of classic software design tradeoffs involving inheritance and container classes, and resolving it is beyond the scope of this document.
+If we do use different types, one of the most important operations on the "executed" forms will of course be transformation to a "predicted" form for reproduction.
```

Comment:
Is there a planned interface for querying dataset provenance?  For example:

- Post-processing a config was deemed to have "bad" value(s).  What runs or do get more precise what data products were produced using this bad config?  Similar question for a "bad" raw or calibration file.  (Doesn't have to be bad, it could be an interesting feature that want to see in different coadds, for example)
- Given a coadd, what raws (or other dataset type in between) affected the coadd.
- Want to make sure particular datasets aren't relevant to a data release before purging.